### ISSUE TITLE
Host release candidates in the BCR to address release candence mismatch

### ISSUE THREAD

**Issue by tpudlik** [bazelbuild/bazel#23461]
### Description of the feature request:

The FR is for the BCR to allow uploading release candidate versions to the registry, which might be git repositories (rather than HTTP archives).

So really there are two related items:

1. Allow release candidate versions. This is mostly a policy decision. We should double-check that the [SemVer syntax for pre-release versions](https://semver.org/#spec-item-9) is supported and correctly interpreted (i.e., `2.0.0-alpha` is a higher version than `1.0.0` but a lower version than `2.0.0`), though [the documentation](https://bazel.build/external/module#version_format) implies this is already the case.
2. Allow release candidate versions to be git repositories rather than HTTP archives.

It's possible that we might want special handling of release candidate versions (e.g., warn the user if version resolution bumps them from a stable version of a transitive dependency to a release candidate; skip release candidates when incrementing versions if the corresponding released version is also in the BCR, i.e. prefer going from `1.0.0` straight to `2.0.0` if it's available, rather than `2.0.0-alpha`). To support this, release candidate versions might need special metadata.

@Wyverald 

### Which category does this issue belong to?

External Dependency

### What underlying problem are you trying to solve with this feature?

Pigweed (a Bazel module) depends on the pico-sdk (another Bazel module), but has a higher release cadence. We effectively release Pigweed continuously, while the pico-sdk has historically been released only once every few months.

This leads to issues when a bug affecting Pigweed is discovered in the pico-sdk. Even once the fix is merged into the upstream pico-sdk, we can't rely on the fix in Pigweed until a release is cut. For local development, we can use a `git_override`, but there's no way for downstream users of Pigweed to inherit this when they specify a version of Pigweed that needs the fix as a `bazel_dep`. This means we need to go beyond the BCR for managing dependencies (e.g. through maintaining our own registry).

To continue expressing our dependencies exclusively via the BCR, we'd like to be able to say that we depend on the release candidate (even though the release has not been cut yet, and there's no HTTP archive). Note that this is different from patching the pico-sdk in the BCR. We don't want to patch it. We just want to make the BCR versions more granular than the GitHub releases.

Another dependency for which we'd like to do something similar is `rules_cc`. There's [a release candidate](https://github.com/bazelbuild/rules_cc/releases/tag/0.0.10-rc1) for `rules_cc` (actually a HTTP archive!) that introduces a new API. But it's not officially released yet, and so not in the BCR.

The reality of software development is that different projects have wildly different release cadences, and so sometimes modules need to depend on yet-unreleased versions of other modules. It should be possible to express this through the BCR.

As Bazel is adopted by large, multi-build system projects, being able to decouple BCR versions from official releases in this way will become increasingly important. These projects will not always be releasing frequently enough for the ecosystem to rely exclusively on released versions, and their maintainers may be unwilling to cut releases just to make their (relatively few) Bazel users' dependency management tidier.

### Which operating system are you running Bazel on?

Any

### Any other information, logs, or outputs that you want to share?

See https://issues.pigweed.dev/issues/360950550#comment3 and following for further context. Basically, this FR attempts to propose a long-term solution for managing unreleased dependencies that falls short of setting up separate registries.

**Comment by fmeum** [comment]
I think that if we want to add support for this, we should also consider the "single-module registry" idea that has been floated around before. Assuming that a Git repo contains a version of a Bazel module with other versions available in the BCR, we could conceivably allow deps such as:
`bazel_dep(name = "github.com/org/repo", version = "v1.2.3-20191109021931-daa7c04131f5")` (inspired by Go's pseudo-version).

Such a dep would not need to count as an override as its version is amenable to relaxed semver comparison (the format needs to be slightly more complex if the base tag itself uses relaxed semver or `-`).

The advantage is that projects could choose to depend on deps that aren't in the BCR yet at all. Hopefully the more cumbersome dep statement would still encourage users to add released versions to the BCR.

**Comment by armandomontanez** [comment]
Related: I started a discussion about how we should name these versions here: https://github.com/bazelbuild/bazel-central-registry/discussions/2888

**Comment by armandomontanez** [comment]
Seems to break pretty early on since there's no source archive URL to grab out:

https://github.com/bazelbuild/bazel-central-registry/pull/3015

**Comment by armandomontanez** [comment]
https://github.com/bazelbuild/bazel-central-registry/pull/3580 is the first BCR entry to use a `git_repository` source. 

I originally wanted to use `.bcr.20250113215903+4b6e6475` (time down to the second of the commit merge for more granularity) as the suffix but hit two issues:
1. Java crashes with large integers. Had to shorten the timestamp to be YYYYmmdd.
2. Everything after the plus (`+4b6e6475`) was ignored in a way where the expected version source string was `*.bcr.20250113`. I didn't want to debug this, so I switched to using a dash (`-`), which has the unfortunate side effect of allowing the commit hash to participate in version resolution.

**Comment by meteorcloudy** [comment]
I believe `+` can no longer appear in version due to https://github.com/bazelbuild/bazel/issues/23127

**Comment by jwnimmer-tri** [comment]
>  You can submit rc modules to BCR with git_repository in source.json ...
>
> https://github.com/bazelbuild/bazel-central-registry/pull/3580 is the first BCR entry to use a git_repository source.

Indeed, to this day `@pico-sdk` is the _only_ module in BCR to use `"type": "git_repository"`.

The downside of allowing `git_repository` in BCR is that it degrades BCR's artifact mirroring ecosystem.  For example, the `.bazelrc` setting `common --module_mirrors=https://bcr.cloudflaremirrors.com` automatically provides backup of https urls in case github is having an outage, but that doesn't help modules whose sole source is from a git server.

In contrast, a `git_repository` allows only a single `remote` to be specified.

Since this ticket was originally discussed, there are new conventions taking shape.

For actual *release candidates* on GitHub, they can be added just like any other release, e.g., https://github.com/bazelbuild/bazel-central-registry/pull/6379 added a `rules_cc@0.2.14-rc1` release candidate, with a stable URL, etc.  They can be `"type": "archive"` like everything else.

For untagged development snapshots, they can be added using a GitHub https `/archive/{commit_sha}.zip` URL, e.g., https://github.com/bazelbuild/bazel-central-registry/pull/8279 added a `googleapis@0.0.0-20260402-c8ca5bce` snapshot.  This requires the "skip url stability check", but with widespread mirroring now (such as the cloudflare mirror), I would say its acceptable.

Keeping around `git_repository` as an option in `source.json` is a nice feature for registries in general (e.g., for in-house registries), but I'd like to float the idea that BCR should disallow it moving forward.  (I can re-post this to a new issue against BCR, if that seems worthwhile.)

**Comment by fmeum** [comment]
Please do file a new issue. It seems reasonable to keep supporting git_repository for custom registries only now that we have module mirrors.
